File: frontend/src/api/streaks.ts
import { api } from "@/api/client";
import type { Author } from "@/api/social";

export interface StreaksToday {
  day: string;
  gym_checked_in: boolean;
  muscles: string[];
  protein_target_g: number | null;
  protein_logged_g: number;
  protein_met: boolean;
}

/** Muscle groups offered at check-in (must match the backend MUSCLE_GROUPS set). */
export const MUSCLE_GROUPS = [
  "chest", "back", "shoulders", "biceps", "triceps", "forearms",
  "core", "quads", "hamstrings", "glutes", "calves", "cardio",
] as const;

export interface MyStreaks {
  gym_streak: number;
  protein_streak: number;
  today: StreaksToday;
}

export interface CheckinResult {
  gym_checked_in: boolean;
  gym_streak: number;
  muscles: string[];
}

export interface LeaderboardEntry {
  user: Author;
  gym_streak: number;
  protein_streak: number;
  gym_days: number;
  protein_days: number;
  score: number;
  is_me: boolean;
}

export interface Leaderboard {
  window_days: number;
  entries: LeaderboardEntry[];
}

export const getMyStreaks = () => api<MyStreaks>("/streaks/me");

export const gymCheckin = (muscles: string[] = []) =>
  api<CheckinResult>("/streaks/checkin", {
    method: "POST",
    body: JSON.stringify({ muscles }),
  });

export const undoGymCheckin = () => api<CheckinResult>("/streaks/checkin", { method: "DELETE" });

export const getLeaderboard = () => api<Leaderboard>("/streaks/leaderboard");

File: backend/app/schemas/streaks.py
"""Streaks + leaderboard schemas (docs/02 §3.4)."""
from __future__ import annotations

import uuid
from datetime import date

from pydantic import BaseModel, Field

from app.schemas.social import Author


class CheckinCreate(BaseModel):
    """Optional payload for a gym check-in: which muscle groups were trained."""

    muscles: list[str] = Field(default_factory=list, max_length=12)


class StreaksToday(BaseModel):
    """Where today stands for the two habits."""

    day: date
    gym_checked_in: bool
    muscles: list[str] = Field(default_factory=list)  # muscle groups logged today
    protein_target_g: float | None  # None when no bodyweight on file
    protein_logged_g: float
    protein_met: bool


class MyStreaks(BaseModel):
    gym_streak: int
    protein_streak: int
    today: StreaksToday


class CheckinResult(BaseModel):
    gym_checked_in: bool
    gym_streak: int
    muscles: list[str] = Field(default_factory=list)


class LeaderboardEntry(BaseModel):
    user: Author
    gym_streak: int
    protein_streak: int
    gym_days: int       # qualifying gym days in the window
    protein_days: int   # qualifying protein days in the window
    score: int          # gym_days + protein_days (ranking key)
    is_me: bool = False


class Leaderboard(BaseModel):
    window_days: int
    entries: list[LeaderboardEntry]
